runtime/vm/resolvesupport.cpp

```diff
@@ -1979,6 +2100,59 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
 
 }
 
+j9object_t
+resolveMethodHandle(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpIndex, UDATA resolveFlags)
+{
+	J9JavaVM *vm = vmThread->javaVM;
+	j9object_t memberName = NULL;
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+	bool canRunJavaCode = J9_ARE_NO_BITS_SET(resolveFlags, J9_RESOLVE_FLAG_JIT_COMPILE_TIME | J9_RESOLVE_FLAG_REDEFINE_CLASS);
+
+	if (!canRunJavaCode) {
+		return NULL;
+	}
+
+	J9RAMMethodRef *ramCPEntry = (J9RAMMethodRef*)ramCP + cpIndex;
+	UDATA invokeCacheIndex = ramCPEntry->methodIndexAndArgCount >> 8;
+	J9Class *ramClass = J9_CLASS_FROM_CP(ramCP);
+	J9ROMMethodRef *romMethodRef = (J9ROMMethodRef *)&ramCP->romConstantPool[cpIndex];
+	J9ROMNameAndSignature *nameAndSig = J9ROMMETHODREF_NAMEANDSIGNATURE(romMethodRef);
+
+	/* Resolve the class. */
+	J9Class *resolvedClass = resolveClassRef(vmThread, ramCP, romMethodRef->classRefCPIndex, resolveFlags);
+	if (resolvedClass != NULL) {
+		/* Create appendix array*/
+		J9Class *objectClass = J9VMJAVALANGOBJECT(vm);
+		J9Class *arrayClass = objectClass->arrayClass;
+		if (NULL == arrayClass) {
+			/* Allocate an array class */
+			J9ROMArrayClass* arrayOfObjectsROMClass = (J9ROMArrayClass*)J9ROMIMAGEHEADER_FIRSTCLASS(vmThread->javaVM->arrayROMClasses);
+			arrayClass = vmThread->javaVM->internalVMFunctions->internalCreateArrayClass(vmThread, arrayOfObjectsROMClass, objectClass);
+		}
+		j9object_t appendix = vm->memoryManagerFunctions->J9AllocateIndexableObject(vmThread, arrayClass, 1, J9_GC_ALLOCATE_OBJECT_INSTRUMENTABLE);
+
+		sendResolveMethodHandle2(vmThread, ramCP, cpIndex, MH_REF_INVOKEVIRTUAL, resolvedClass, nameAndSig, appendix);
+		memberName = (j9object_t)vmThread->returnValue;
+
+		if (vmThread->currentException != NULL) {
+			/* Already a pending exception */
+			memberName = NULL;
+		} else if (memberName == NULL) {
+			setCurrentExceptionUTF(vmThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
+		}
+
+		if (memberName != NULL) {
+			/* store result */
+			J9InvokeCacheEntry *resultEntry = ((J9InvokeCacheEntry *)ramClass->invokeCache) + invokeCacheIndex;
+			resultEntry->target = memberName;
+			resultEntry->appendix = (j9object_t)J9JAVAARRAYOFOBJECT_LOAD(vmThread, appendix, 0);
```

Comment:
What kind of object is appendix?  And what's the relationship to the memberName?  I want to ensure that racing threads overwriting each other here won't lead to inconsistent data problems

Reply:
How do we guarantee that anyone checking the resolved status sees both slots at the same time?  Power has a weak memory order which means the writes may be seen out of order or only 1 may be seen on other threads.

If `target` is used for the resolve check, we need to ensure that no-one can read only 1 of the values implying we either need to write appendix first, and then barrier the reads,
OR
check both fields to determine if the entry is resolved or not